### DISCUSSION TITLE
Start bit

### DISCUSSION THREAD

**Discussion by superaga** [cantools/cantools#571] [Q&A] [answered]
Hi all,
as I consider _cantools_ a very mature library I don't understand what I'm missing to make it properly "work", more specifically I want to "define" which bit position the _Message_ class should use to identify the _start bit_.
Sorry to referencing here _canmatrix_, but in [its documentation](https://github.com/ebroecker/canmatrix/wiki/signal-Byteorder) it is (very well) stated that the _start bit_ reference it is independent of the endianness of the signal ("_Usually the start bit referenced by the file formats reference the start of the signal data within the message independent of the endianess of the signal. With some formats (DBF), the startbit may always reference the LSB of the signal data._").

Sorry but to me looks that _cantools_ wants the same "standard" the raw dbc is using, but this is "wrong".

Vector, which created the dbc, defined in the raw file the _start bit_ depending by the endinness but ...
CANDB++ (as I read also here is considered the "reference" program) shows the signals referencing the _start bit_ to be always in the LSB position. 

So with a dbc raw file like this:

```
BO_ 1 MSB_Message: 8 Vector__XXX
 SG_ MSB_signal_long : 7|12@0+ (1,0) [0|0] "" Vector__XXX
 SG_ MSB_signal_short : 11|4@0+ (1,0) [0|0] "" Vector__XXX

BO_ 2 LSB_Message: 8 Vector__XXX
 SG_ LSB_signal_long : 0|12@1+ (1,0) [0|0] "" Vector__XXX
 SG_ LSB_signal_shoer : 12|4@1+ (1,0) [0|0] "" Vector__XXX
```
The _start bit_ of the MSB signal (_MSB_signal_long_) appears to be in position 7, while it should be in position 12 (as it is in CANDB++).
The _start bit_ of the LSB signal (_LSB_signal_long_) appears to be in position 0, which is correct for both the raw file and the CANDB++.

Now, independently from the "correct/wrong" way the raw dbc represent the signals, I think that cantools should accept this info in the "correct" form i.e. _start bit_ in the LSB position, but when I define the message (_MSB_Message_) as above:

- MSB_signal_long:
  - start bit: 12
  - length: 12
- MSB_signal_short:
  - start bit: 8
  - length: 4

I get the overlapping message:
_"Exception has occurred: Error The signals MSB_signal_short and MSB_signal_long are overlapping in message MSB_Message."_

As I'm sure that what I defined is correct, I was wondering if _cantools_ would have some flag to specify which behavior use to reference the _start bit_ because unfortunately I didn't find any.
It is eventually possible add it?

Many thanks!

**Comment by andlaus** [comment]
> Sorry but to me looks that cantools wants the same "standard" the raw dbc is using, but this is "wrong".

depends on what you mean by "wrong": if you mean that it can make your brain hurt, I agree, but it is not "incorrect" in an absolute sense. I'm not sure how canmatrix handles this, but given that the layout of signals crossing byte boundaries is completely different depending on the endianness (see [here](https://github.com/cantools/cantools/blob/master/cantools/database/can/signal.py#L97-L115)), if you demand that a signal in a given bit ordering always occupies the bits [start_bit, start_bit+bit_length],  I cannot see how to unify this...

**Reply by superaga** [reply]
Hi @andlaus,

no, it is not because my brain 😅, the reason it is because _Simulink_, which receives the dbc works in the opposite way, from here the "need" to keep the things aligned across the entire project.
I know what you mean, you when you say that _"it is not "incorrect" in an absolute sense"_, I agree with you... The problem is that _Vector_ never defined a standard as it is a proprietary incoherent format. 
I'll update my script in order to provide to _cantools_ the right start bit.

Do you think that would be useful to open a request to have a flag that specify if the start bit should be in the lsb or in the msb position?
Many thanks!

**Comment by andlaus** [comment]
> Do you think that would be useful to open a request to have a flag that specify if the start bit should be in the lsb or in the msb position?

I don't think so: there already is a [`sawtooth_to_network_bitnum()`](https://github.com/cantools/cantools/blob/master/cantools/database/utils.py#L284) utility function which converts these, so if I understood you correctly, you want something analogous to:
```python
if signal.byte_order == 'big_endian':
  pos = sawtooth_to_network_bitnum(pos)
```
correct? (what I personally find terribly confusing in this context, is that switching the "byte" order seems to also reverse the bit ordering...)

**Reply by superaga** [reply]
I just checked this function and doesn't change the start bit to the lsb position. Also the function name suggests something different, but I didn't investigate too much what it does.
I made my function to fix this, if anyone is interested here the code:

```
def swap_start_bit(self, start_bit, length) -> int:
        abs_byte_pos, rel_bit_pos = divmod(start_bit,8)
        pos_to_bit_7 = 7 - rel_bit_pos
        if length <= (pos_to_bit_7 + 1):
            return start_bit + length - 1
        else:
            # identify the start bit position
            row, col = divmod(start_bit - length, 8)
            # find the new start bit
            if start_bit - length == 0:
                lsb = 7
            else:
                pos_bit_0 = (row + 1) * 8 
                lsb = pos_bit_0 + 7 - col
            return lsb
```

**Comment by superaga** [comment, marked as answer]
Yesterday I come across another issue with the previous code, so I finally decided to fix it once and for all.
This is the general formulation to convert the start bit from the big endian to little endian, the one used by cantools.


```
    def start_bit_MSB_to_LSB(self, lsb, length) -> int:
        # DBC matrix size
        max_row = 8
        max_col = 8
        # create an the bit 7 value for each DBC matrix row
        bit_7 = [row for row in range(7, 64, max_row)]
        # create and unroll the DBC matrix
        bit_position = [bit for row in bit_7 for bit in range(row, row - max_col, -1)]
        # find the current lsb bit within the DBC matrix
        lsb_position = bit_position.index(lsb)
        # find the msb position within the DBC matrix
        msb_position = lsb_position - length + 1
        # extract the msb bit from the DBC matrix
        msb = bit_position[msb_position]
        return msb
```
